src/Datadog.Trace/Propagation/ContextPropagatorBuilder.cs

```diff
@@ -7,22 +7,31 @@ namespace Datadog.Trace.Propagation
 {
     internal static class ContextPropagatorBuilder
     {
-        private static readonly IReadOnlyDictionary<PropagatorType, Func<ITraceIdConvention, IPropagator>> _propagatorSelector =
+        private static readonly IReadOnlyDictionary<PropagatorType, Func<ITraceIdConvention, IPropagator>> PropagatorSelector =
             new Dictionary<PropagatorType, Func<ITraceIdConvention, IPropagator>>()
             {
-                { PropagatorType.B3, (convention) => new B3SpanContextPropagator(convention) },
-                { PropagatorType.Datadog, (convention) => new DDSpanContextPropagator(convention) },
-                { PropagatorType.Default, (convention) => new DDSpanContextPropagator(convention) },
+                { PropagatorType.W3C, convention => new W3CSpanContextPropagator(convention) },
+                { PropagatorType.B3, convention => new B3SpanContextPropagator(convention) },
+                { PropagatorType.Datadog, convention => new DDSpanContextPropagator(convention) },
+                { PropagatorType.Default, convention => new DDSpanContextPropagator(convention) },
             };
 
         public static IPropagator BuildPropagator(PropagatorType propagator, ITraceIdConvention traceIdConvention)
         {
-            if (_propagatorSelector.TryGetValue(propagator, out Func<ITraceIdConvention, IPropagator> getter))
+            if (PropagatorSelector.TryGetValue(propagator, out Func<ITraceIdConvention, IPropagator> getter))
             {
-                return getter(traceIdConvention);
+                // W3C propagator requires Otel TraceId convention as it's specification clearly states lengths of traceId and spanId values in the header.
+                return IsW3CButNotOtelTraceId(propagator, traceIdConvention)
```

Comment:
I would throw an unsupported exception here. Cause if config is asking for W3C and gets default, it might cause more misunderstanding.

Reply:
Well, that actually sounds reasonable 👍